packages/contracts/src/contracts/current/forwarder/MixinBuyExactAssets.sol

```diff
@@ -0,0 +1,137 @@
+    ///      This function is payable and will convert all incoming ETH into WETH and perform the trade on behalf of the caller.
+    ///      The caller is sent all assets from the fill of orders. This function will revert unless the requested amount of assets are purchased.
+    ///      Any excess ETH sent will be returned to the caller
+    /// @param orders An array of Order struct containing order specifications.
+    /// @param signatures An array of Proof that order has been created by maker.
+    /// @param feeOrders An array of Order struct containing order specifications for fees.
+    /// @param assetAmount The amount of maker asset to buy.
+    /// @param feeSignatures An array of Proof that order has been created by maker for the fee orders.
+    /// @param feeProportion A proportion deducted off the ETH spent and sent to feeRecipient. The maximum value for this
+    ///        is 1000, aka 10%. Supports up to 2 decimal places. I.e 0.59% is 59.
+    /// @param feeRecipient An address of the fee recipient whom receives feeProportion of ETH.
+    /// @return FillResults amounts filled and fees paid by maker and taker.
+    function buyExactAssets(
+        Order[] memory orders,
+        bytes[] memory signatures,
+        Order[] memory feeOrders,
+        bytes[] memory feeSignatures,
+        uint256 assetAmount,
+        uint16  feeProportion,
+        address feeRecipient)
+        payable
+        public
+        returns (Exchange.FillResults memory totalFillResults)
+    {
+        require(msg.value > 0, VALUE_GREATER_THAN_ZERO);
+        address takerAsset = readAddress(orders[0].takerAssetData, 0);
+        require(takerAsset == address(ETHER_TOKEN), TAKER_ASSET_WETH_REQUIRED);
+        uint8 proxyId = uint8(orders[0].makerAssetData[orders[0].makerAssetData.length - 1]);
+        require(proxyId == ERC20_PROXY_ID || proxyId == ERC721_PROXY_ID, UNSUPPORTED_TOKEN_PROXY);
+
+        uint256 remainingTakerAssetAmount = msg.value;
+        ETHER_TOKEN.deposit.value(remainingTakerAssetAmount)();
+        if (proxyId == ERC20_PROXY_ID) {
+            totalFillResults = buyExactERC20TokensInternal(orders, signatures, feeOrders, feeSignatures, assetAmount);
+        } else if (proxyId == ERC721_PROXY_ID) {
+            totalFillResults = buyExactERC721TokensInternal(orders, signatures, feeOrders, feeSignatures, assetAmount);
+        }
+        remainingTakerAssetAmount = safeSub(remainingTakerAssetAmount, totalFillResults.takerAssetFilledAmount);
+        // Prevent a user from paying too much in fees through fee abstraction
+        require(
+            isAcceptableThreshold(
+                remainingTakerAssetAmount,
+                totalFillResults.takerAssetFilledAmount),
+            UNACCEPTABLE_THRESHOLD);
+        withdrawPayAndDeductFee(remainingTakerAssetAmount, totalFillResults.takerAssetFilledAmount, feeProportion, feeRecipient);
+        return totalFillResults;
+    }
+
+    function buyExactERC20TokensInternal(
```

Comment:
We've been adding function descriptions to the protocol contracts, and I think it could be useful in the forwarding contract as well. I realize we started doing this after this PR was created =\, but it would be super helpful.
Ex for this function:
```
/// @dev Buys an exact amount of an ERC20 token using WETH.
/// @param orders Orders to fill. The maker asset is the ERC20 token to buy. The taker asset is WETH.
/// @param signatures Proof that the orders were created by their respective makers.
/// @param feeOrders to fill. The maker asset is ZRX and the taker asset is WETH.
/// @param feeSignatures Proof that the feeOrders were created by their respective makers.
/// @param assetAmount Amount of the ERC20 token to buy.
/// @return totalFillResults Aggregated fill results of buying the ERC20 and ZRX tokens.
```